src/cfg/domtree.h

```diff
@@ -0,0 +1,178 @@
+  //    the dominator tree, and that is our public interface, so name it
+  //    explicitly as the immediate dominators.
+  //  * Indexes are reversed. The paper uses postorder indexes, i.e., the entry
+  //    node has the highest index, while we have a natural indexing since we
+  //    traverse in reverse postorder anyhow (see cfg-traversal.h), that, is,
+  //    the entry has the lowest index.
+  //  * finger1, finger2 => left, right.
+  //  * We assume the input is reducible, since wasm and Binaryen IR have that
+  //    property. This simplifies some things, see below.
+  //
+  // Otherwise this is basically a direct implementation.
+  //
+  // [1] Cooper, Keith D.; Harvey, Timothy J; Kennedy, Ken (2001). "A Simple,
+  //       Fast Dominance Algorithm" (PDF).
+  //       http://www.hipersoft.rice.edu/grads/publications/dom14.pdf
+
+  // If there are no blocks, we have nothing to do.
+  Index numBlocks = blocks.size();
+  if (numBlocks == 0) {
+    return;
+  }
+
+  // Map basic blocks to their indices.
+  std::unordered_map<BasicBlock*, Index> blockIndices;
+  for (Index i = 0; i < numBlocks; i++) {
+    blockIndices[blocks[i].get()] = i;
+  }
+
+  // Use a nonsense value to indicate what has yet to be initialized.
+  const Index nonsense = -1;
+
+  // Initialize the iDoms array. The entry starts with its own index, which is
+  // used as a guard value in effect (we will never process it, and we will fix
+  // up this value at the very end). All other nodes start with a nonsense value
+  // that indicates they have yet to be processed.
+  iDoms.resize(numBlocks, nonsense);
+  iDoms[0] = 0;
+
+  // Process the (non-entry) blocks in reverse postorder, computing the
+  // immediate dominators as we go. This returns whether we made any changes,
+  // which is used in an assertion later down.
+  auto processBlocks = [&]() {
+    bool changed = false;
+    for (Index index = 1; index < numBlocks; index++) {
+      // Loop over the predecessors. Our new parent is basically the
+      // intersection of all of theirs: our immediate dominator must precede all
+      // of them.
+      auto& preds = blocks[index]->in;
+      Index newParent = nonsense;
+      for (auto* pred : preds) {
+        auto predIndex = blockIndices[pred];
+
+        // In a reducible graph, we only need to care about the predecessors
+        // that appear before us in the reverse postorder numbering. The only
+        // predecessor that can appear *after* us is a loop backedge, but that
+        // will never dominate the loop - the loop is dominated by its single
+        // entry (since it is reducible, it has just one entry).
+        if (predIndex > index) {
+          continue;
+        }
+
+        // All of our predecessors will have been processed before us, except
+        // if they are unreachable from the entry, in which case, we can ignore
+        // them.
+        if (iDoms[predIndex] == nonsense) {
+          continue;
+        }
```

Comment:
Are unreachable blocks possible to be generated from Binaryen AST?

Reply:
Yes, e.g.
```wat
(unreachable)
(nop)
(if
 ..
..
)
(nop)
```
The basic block after the unreachable has no predecessor. And the diamond pattern after it are also effectively unreachable, and they are the reason for that commit + test.

CFGWalker atm does not try to avoid creating unreachable blocks (I guess it could, but it would need to look at all predecessors etc.).